File: sem2/po/PO-VirtualWorld-Python/simulator/organisms/plants/Plant.py
import random

from simulator.organisms.Organism import Organism

SEEDING_CHANCE_BASE = 100
BASE_PLANT_SEEDING_THRESHOLD = 97
BASE_PLANT_STRENGTH = 0
BASE_PLANT_INITIATIVE = 0


class Plant(Organism):
    def __init__(self, new_world, new_location, new_strength, new_initiative, new_type, new_seeding_threshold):
        super().__init__(new_world, new_location, new_strength, new_initiative, new_type)
        self.seeding_threshold = new_seeding_threshold

    def action(self):
        self.seed()

    def collision(self, invader):
        invader.update_location(self.location)
        self.kill_self()

    def seed(self):
        r = random.randint(0, SEEDING_CHANCE_BASE)

        if r > self.seeding_threshold:
            free_loc = self.world.map.get_first_free_neighbour(self.location)
            if free_loc is not None:
                self.tlog(" seeded at " + str(free_loc))
                self.world.add_organism(self.spawn_offspring(free_loc))

    def __str__(self):
        return "Plant " + str(self.location)

File: sem2/po/po_proj1/Plants/Plant.cpp
//
// Created by M Sz on 05/04/2023.
//

#include "Plant.h"

Plant::Plant(int new_strength, int new_x, int new_y, char new_symbol,int seeding_chance, World &new_world)
        : Organism(new_strength, BASE_PLANT_INITIATIVE, new_x, new_y, new_symbol, new_world), seeding_threshold(seeding_chance)
{}

Plant::Plant(int new_strength, XY new_map_location, char new_symbol,int seeding_chance, World &new_world)
        : Organism(new_strength, BASE_PLANT_INITIATIVE, new_map_location, new_symbol, new_world), seeding_threshold(seeding_chance)
{}

void Plant::action() {
    this->seed();
}

void Plant::seed() {
    if (rand() % SEEDING_CHANCE_BASE < this->get_seeding_threshold()) {
        //this->log(PLANT_IDLE_MESSAGE);
        return;
    }


    
    for (int i = -1; i < 2; ++i) {
        for (int j = -1; j < 2; ++j) {
            if (i == 0 && j == 0)
                continue;

            if (!this->get_world().get_map().is_out_of_map(this->get_x() + i, this->get_y() + j)) {
                if (this->get_world().get_organism_at_location(this->get_x() + i, this->get_y() + j) == nullptr) {

                    this->get_world().add_organism(this->seed_new_plant(this->get_x() + i, this->get_y() + j));

                    std::string s(SEEDING_MESSAGE);
                    s += this->login();
                    s += " (";
                    s += std::to_string(this->get_x() + i);
                    s += ", ";
                    s += std::to_string(this->get_y() + j);
                    s += ")";

                    this->log(s);

                    return;
                }
            }
        }
    }
}

void Plant::set_seeding_threshold(int new_seeding_threshold) {
    this->seeding_threshold = new_seeding_threshold;
}

int Plant::get_seeding_threshold() const {
    return this->seeding_threshold;
}

void Plant::collision(Organism *invading_organism) {
    invading_organism->update_location(this->get_location());
    this->kill_self();
    }

std::string Plant::login() {
    return std::string("Plant");
}




